doc/threadsafe_function_ex.md

```diff
@@ -0,0 +1,243 @@
+
+Creates a new instance of the `Napi::ThreadSafeFunctionEx` object.
+
+```cpp
+Napi::ThreadSafeFunctionEx<ContextType, DataType, Callback>::ThreadSafeFunctionEx(napi_threadsafe_function tsfn);
+```
+
+- `tsfn`: The `napi_threadsafe_function` which is a handle for an existing
+  thread-safe function.
+
+Returns a non-empty `Napi::ThreadSafeFunctionEx` instance. To ensure the API
+statically handles the correct return type for `GetContext()` and
+`[Non]BlockingCall()`, pass the proper type arguments to
+`Napi::ThreadSafeFunctionEx`.
+
+### New
+
+Creates a new instance of the `Napi::ThreadSafeFunctionEx` object. The `New`
+function has several overloads for the various optional parameters: skip the
+optional parameter for that specific overload.
+
+```cpp
+New(napi_env env,
+    CallbackType callback,
+    const Object& resource,
+    ResourceString resourceName,
+    size_t maxQueueSize,
+    size_t initialThreadCount,
+    ContextType* context,
+    Finalizer finalizeCallback,
+    FinalizerDataType* data = nullptr);
+```
+
+- `env`: The `napi_env` environment in which to construct the
+  `Napi::ThreadSafeFunction` object.
+- `[optional] callback`: The `Function` to call from another thread.
+- `[optional] resource`: An object associated with the async work that will be
+  passed to possible async_hooks init hooks.
+- `resourceName`: A JavaScript string to provide an identifier for the kind of
+  resource that is being provided for diagnostic information exposed by the
+  async_hooks API.
+- `maxQueueSize`: Maximum size of the queue. `0` for no limit.
+- `initialThreadCount`: The initial number of threads, including the main
+  thread, which will be making use of this function.
+- `[optional] context`: Data to attach to the resulting `ThreadSafeFunction`.
+  Can be retreived via `GetContext()`.
+- `[optional] finalizeCallback`: Function to call when the
+  `ThreadSafeFunctionEx` is being destroyed.  This callback will be invoked on
+  the main thread when the thread-safe function is about to be destroyed. It
+  receives the context and the finalize data given during construction (if
+  given), and provides an opportunity for cleaning up after the threads e.g. by
+  calling `uv_thread_join()`. It is important that, aside from the main loop
+  thread, there be no threads left using the thread-safe function after the
+  finalize callback completes. Must implement `void operator()(Env env,
+  DataType* data, ContextType* hint)`.
+- `[optional] data`: Data to be passed to `finalizeCallback`.
+
+Returns a non-empty `Napi::ThreadSafeFunctionEx` instance.
+
+Depending on the targetted `NAPI_VERSION`, the API has different implementations
+for `CallbackType callback`.
+
+When targetting version 4, `CallbackType` is:
+- `const Function&`
+- skipped, in which case the API creates a new no-op `Function`
+
+When targetting version 5+, `CallbackType` is:
+- `const Function&`
+- `std::nullptr_t`
+- skipped, in which case the API passes `std::nullptr`
```

Comment:
This is unclear, especially how the word "skipped" fits in. Is this supposed to mean "When targetting version <x> and `CallbackType` is skipped..."?

Reply:
This has been addressed [here](https://github.com/nodejs/node-addon-api/pull/742/files#diff-3c83591b72a5c006a3ff969a41d5dc61R93-R101).